src/file_utils.cpp

```diff
@@ -1,65 +1,42 @@
 #include "openmc/file_utils.h"
 
-#include <algorithm> // any_of
-#include <cctype>    // for isalpha
-#include <fstream>   // for ifstream
-#include <sys/stat.h>
+#include <filesystem>
 
 namespace openmc {
 
 bool dir_exists(const std::string& path)
 {
-  struct stat s;
-  if (stat(path.c_str(), &s) != 0)
-    return false;
-
-  return s.st_mode & S_IFDIR;
+  std::filesystem::path d(path);
+  return std::filesystem::is_directory(d);
 }
 
 bool file_exists(const std::string& filename)
 {
-  // rule out file being a path to a directory
-  if (dir_exists(filename))
+  std::filesystem::path p(filename);
+  if (!std::filesystem::exists(p)) {
     return false;
-
-  std::ifstream s {filename};
-  return s.good();
+  }
+  if (std::filesystem::is_directory(p)) {
+    return false;
+  }
+  return true;
 }
 
 std::string dir_name(const std::string& filename)
 {
-  size_t pos = filename.find_last_of("\\/");
-  return (std::string::npos == pos) ? "" : filename.substr(0, pos + 1);
+  std::filesystem::path p(filename);
+  return (p.parent_path()).string();
 }
 
 std::string get_file_extension(const std::string& filename)
 {
-  // try our best to work on windows...
-#if defined(_WIN32) || defined(_WIN64)
-  const char sep_char = '\\';
-#else
-  const char sep_char = '/';
-#endif
-
-  // check that at least one letter is present
-  const auto last_period_pos = filename.find_last_of('.');
-  const auto last_sep_pos = filename.find_last_of(sep_char);
-
-  // no file extension. In the first case, we are only given
-  // a file name. In the second, we have been given a file path.
-  // If that's the case, periods are allowed in directory names,
-  // but have the interpretation as preceding a file extension
-  // after the last separator.
-  if (last_period_pos == std::string::npos ||
-      (last_sep_pos < std::string::npos && last_period_pos < last_sep_pos))
-    return "";
-
-  const std::string ending = filename.substr(last_period_pos + 1);
-
-  // check that at least one character is present.
-  const bool has_alpha = std::any_of(ending.begin(), ending.end(),
-    [](char x) { return static_cast<bool>(std::isalpha(x)); });
-  return has_alpha ? ending : "";
```

Comment:
In debugging why the source point regression tests were failing, I found this change.

In calls to `write_source_point`, we both don't add the `.h5` extension, but also add the batch number to the end of the file

https://github.com/openmc-dev/openmc/blob/3bedd043d0026490412018685901c9409635713b/src/simulation.cpp#L438-L449

This leads to passing something like `source.10` into `get_file_extension`. Previously, openmc only considered something to be an extension if it contained one *alpha* character, not just one character as this comment implies. So `get_file_extension("source.10")` would produce `""`. 

However, `std::filesystem::path::get_extension` only considers the extension to be any text after (and including) the rightmost period., or `".10"`. This was triggered after seeing messages [like](https://github.com/openmc-dev/openmc/actions/runs/9570450045/job/26385224028?pr=3042#step:11:1152)
```
 Creating state point statepoint.10.h5...
 WARNING: write_source_point was passed a file extension differing from .h5, but
          an hdf5 file will be written.
 WARNING: write_source_point was passed a file extension differing from .h5, but
          an hdf5 file will be written.
```
We explicitly add the `.h5` to the *statepoint* file

https://github.com/openmc-dev/openmc/blob/3bedd043d0026490412018685901c9409635713b/src/state_point.cpp#L47-L50

but not the source point files.

I'm of two minds.

1. Ensure that the `.h5` extension is added to `filename`-s given to `write_source_point`, or
2. If `filename` given to `write_source_point` does not have a `.h5` extension, append `.h5`.

1 is my preference since we would be more explicit: we tell you to write a file, write it here. But we already had a pattern of changing the destination by appending `.h5` if the extension didn't match what we liked.

Reply:
I agree -- explicit is better than implicit